### ISSUE TITLE
Efficient computation of NCE scores for f1->f5 prediction

### ISSUE THREAD

**Issue by Philip-Bachman** [mila-iqia/spr#9]
The structure of the InfoNCE computation is pretty inefficient...

https://github.com/ankeshanand/abstract-world-models/blob/9d4a94bf072fe5c1307a7d0f1841f1441cda837f/src/stdim.py#L72

It should be more like:

<pre><code>
def nce_with_negs_from_same_loc(f_glb, f_lcl):
    '''
    Compute InfoNCE cost with source features in f_glb and target features in
    f_lcl. We assume one source feature vector per item in batch and n_locs
    target feature vectors per item in batch. There are n_batch items in the
    batch and the dimension of source/target feature vectors is n_rkhs.
    -- defn: we condition on source features and predict target features

    For the positive nce pair (f_glb[i, :], f_lcl[i, :, l]), which comes from
    batch item i at spatial location l, we will use the target feature vectors
    f_lcl[j, :, l] as negative samples, for all j != i.

    Input:
      f_glb : (n_batch, n_rkhs)          -- one source vector per item
      f_lcl : (n_batch, n_rkhs, n_locs)  -- n_locs target vectors per item
    Output:
      loss_nce : (n_batch, n_locs)       -- InfoNCE cost at each location
    '''
    n_batch = f_glb.size(0)
    n_rkhs = f_glb.size(1)
    n_locs = f_lcl.size(2)
    # reshaping for big matrix multiply
    f_glb = f_glb.permute(1, 0)                                # (n_rkhs, n_batch)
    f_lcl = f_lcl.permute(0, 2, 1)                             # (n_batch, n_locs, n_rkhs)
    f_lcl = f_lcl.reshape(n_batch * n_locs, n_rkhs)            # (n_batch*n_locs, n_rkhs)
    # compute raw scores dot(f_glb[i, :], f_lcl[j, :, l]) for all i, j, l
    raw_scores = torch.mm(f_lcl, f_glb)                        # (n_batch*n_locs, n_batch)
    raw_scores = raw_scores.reshape(n_batch, n_locs, n_batch)  # (n_batch, n_locs, n_batch)
    # now, raw_scores[j, l, i] = dot(f_glb[i, :], f_lcl[j, :, l])
    # -- we can get NCE log softmax by normalizing over the j dimension...
    nce_lsmax = -F.log_softmax(raw_scores, dim=0)              # (n_batch, n_locs, n_batch)
    # make a mask for picking out the log softmax values for positive pairs
    pos_mask = torch.eye(n_batch, dtype=nce_lsmax.dtype, device=nce_lsmax.device)
    pos_mask = pos_mask.unsqueeze(dim=1)
    # use a masked sum over the j dimension to select positive pair NCE scores
    loss_nce = (nce_lsmax * pos_mask).sum(dim=0)               # (n_locs, n_batch)
    # permute axes to make return shape consistent with input shape
    loss_nce = loss_nce.permute(1, 0)                          # (n_batch, n_locs)
    return loss_nce
</code></pre>

(this issue is also present in the code for the Atari ST-DIM paper)
